src/datasets/iterable_dataset.py

```diff
@@ -396,8 +412,12 @@ def map(self, function: Callable, batched: bool = False, batch_size: int = 1000)
         """
         info = copy.deepcopy(self._info)
         info.features = None
+        formatter = _features_formatter(self.features)
         ex_iterable = MappedExamplesIterable(
-            self._ex_iterable, function=function, batched=batched, batch_size=batch_size
+            ((key, formatter(example)) for key, example in self._ex_iterable) if self.features else self._ex_iterable,
```

Comment:
If I'm not mistaken here you are passing an iterator instead of an iterable.
It means that if you iterate over the dataset once, then the second time your iterate over it, it will be empty.
Maybe we can have a TypedExamplesIterable class that does the feature formatting ?

Reply:
Oh, I see. Yes, a new class that does the formatting makes sense.